### PR TITLE
SSI SRV inputs

### PR DESCRIPTION
**Pull request by elvinpoole**
- Added stage to ingest SSI inputs 
- Added stage to compute magnification signal from SSI catalogs

### CODE REVIEW

**Comment by elvinpoole** [comment]
Examples:
`examples/ssi/pipeline_gcr.yml`
Single stage example. Ingests a small SRV SSI runs using GCR and outputs photometry catalog

`examples/ssi/pipeline_mag.yml` 
Runs an example of the magnification pipeline. inputs would typically be two photometry catalogs (one with 0 magnification, one magnified). For this example both catalogs are the example 1deg2 catalog

A more realisatic example using DES data is also ready but will appear in a separate PR

File: txpipe/magnification.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+            #compute mag coeff with the whole sample
+            csample = self.calc_frac_change(w1, w2)/deltak
+            self.write_output(outfile, csample, ibin)
+
+            if self.config["bootstrap_error"]:
+                print(f'Computing uncertainty with bootstrap')
+
+                #compute mag coeff with subsampling patches to get a boostrap error
+                patch1 = cluster.predict(
```
Comment (joezuntz) [inline txpipe/magnification.py:81]: Can you add a comment here explaining what's happening some more - e.g. what kind of object `cluster` is, and what we are `predict`ing.

File: txpipe/magnification.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        # limit number of objects too 1000 per patch so stop the training 
+        # from taking too long  
+        max_per_patch = 1000
+        maxobj = max_per_patch*n_patches
+
+        n_obj_lens = nomag_cat[f'lens/bin_all/ra'].shape[0]
+
+        if n_obj_lens > maxobj:
+            index = np.sort(np.random.choice(n_obj_lens, size=maxobj, replace=False))
+            ra = nomag_cat[f'lens/bin_all/ra'][index]
```
Comment (joezuntz) [inline txpipe/magnification.py:136]: Random access reads like this using HDF can be painfully slow.  It's definitely worth sorting the index before using it, but also leaving a note that if it's too slow then we can look at ways of speeding it up.

File: txpipe/magnification.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        
+        if self.config["bootstrap_error"]:
+            #split no-mag sample up into patches to be used in the bootstrap errors
+            cluster = self.calc_cluster_patches(nomag_cat)
+
+        # compute fractional change in number count for both samples
+        # in each lens bin
+        nbins = nomag_cat['lens/'].attrs['nbin_lens']
+        outfile = self.setup_output(nbins)
+        for ibin in range(nbins):
```
Comment (joezuntz) [inline txpipe/magnification.py:65]: Do we also want to run this on the non-tomographic ("2D") sample?

File: txpipe/magnification.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+                #count number of objects in each patch
+                label1, counts1 = np.unique(patch1, return_counts=True)
+                label2, counts2 = np.unique(patch2, return_counts=True)
+                assert (label1 == np.arange(cluster.n_clusters)).all(), "empty bootstrap patches"
+                assert (label2 == np.arange(cluster.n_clusters)).all(), "empty bootstrap patches"
+
+                # define a function that computes fractional number count change
+                # and takes only patch IDs as input (i.e. fix the counts to this z bin)
+                def calc_frac_change_patches_ibin(patch_ids):
+                    return self.calc_frac_change_patches(patch_ids, counts1, counts2)
```
Comment (joezuntz) [inline txpipe/magnification.py:103]: Is there a reason that the bootstrap part doesn't use the weights but the main calculation does?

**Review by joezuntz — CHANGES_REQUESTED** [review]

File: txpipe/ingest/ssi.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
+from ..base_stage import PipelineStage
+from ..data_types import (
+    YamlFile,
```
[the remark this answers is not in the document]
Comment (elvinpoole) [reply txpipe/ingest/ssi.py]: done

File: txpipe/ingest/ssi.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+        import GCRCatalogs
+
+        output_catalogs = [
+            "injection_catalog",
+            "ssi_photometry_catalog",
+            "ssi_uninjected_photometry_catalog",
+        ]
+
+        for output_catalog_name in output_catalogs:
+            catalog_name = self.config[f"{output_catalog_name}_name"]
```
[the remark this answers is not in the document]
Comment (elvinpoole) [reply txpipe/ingest/ssi.py]: done. also changed `output_catalogs` to `catalog_labels`

File: txpipe/ingest/ssi.py [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+                    qobj = gc0.get_quantities(q)
+                    # TO DO: load with iterator (currently returns full datatset)
+                    # GCRCatalogs.lsst_object.LSSTInjectedObjectTable iterator currently
+                    # returns full dataset
+
+                    if np.isnan(qobj[q]).all():
+                        continue  # skip the quantities that are empty
+
+                    group.create_dataset(q, data=qobj[q], dtype=qobj[q].dtype)
+                    if q in column_names.keys():
```
[the remark this answers is not in the document]
Comment (elvinpoole) [reply txpipe/ingest/ssi.py]: done

File: txpipe/ingest/ssi.py [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+                        # also save with TXPipe names
+                        group.create_dataset(
```
[the remark this answers is not in the document]
Comment (elvinpoole) [reply txpipe/ingest/ssi.py]: done

File: txpipe/ingest/ssi.py [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+                            column_names[q], data=qobj[q], dtype=qobj[q].dtype
+                        )
+
+                except KeyError:  # skip quantities that are missing
+                    warnings.warn(f"KeyError for quantity {q}")
+                    continue
+
+                except TypeError:
+                    warnings.warn(f"TypeError when trying to save quantity {q}")
```
[the remark this answers is not in the document]
Comment (elvinpoole) [reply txpipe/ingest/ssi.py]: done

File: txpipe/ingest/ssi.py [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+        ("ssi_photometry_catalog", HDFFile),
+        ("ssi_uninjected_photometry_catalog", HDFFile),
+    ]
+
+    config_options = {
+        "injection_catalog_name": "",  # Catalog of objects manually injected
+        "ssi_photometry_catalog_name": "",  # Catalog of objects from real data with no injections
+        "ssi_uninjected_photometry_catalog_name": "",  # Catalog of objects from real data with no injections
+        "GCRcatalog_path": "",
+        "flux_name": "gaap3p0Flux",
```
[the remark this answers is not in the document]
Comment (elvinpoole) [reply txpipe/ingest/ssi.py:41]: It was a fairly arbitrary choice. The SRV runs contain gaap fluxes (with a few different apertures) and cmodel fluxes. Cmodel had a lot of nan's so i picked gaap for testing

[the remark this answers is not in the document]
Comment (elvinpoole) [reply txpipe/ingest/ssi.py:41]: maybe we can change the default once it is clear which will be used for a given lens sample selection?

File: txpipe/ingest/ssi.py [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+    }
+
+    def run(self):
+        """
+        Run the analysis for this stage.
+        """
+
+        # loop through chunks of the ssi photometry
+        # catalog and match to the injections
+        matched_cat = self.match_cats()
```
[the remark this answers is not in the document]
Comment (elvinpoole) [reply txpipe/ingest/ssi.py:156]: yes this was for future proofing. I was considering adding alternative matching methods in the future. happy to just put it all in run if you like

File: txpipe/ingest/ssi.py [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+        match_outfile = self.setup_output(
+            "matched_ssi_photometry_catalog",
+            "photometry",
+            inj_cat["photometry"],
+            phot_cat["photometry"],
+            max_n,
+        )
+
+        start1 = 0
+        for ichunk in range(n_chunk):
```
[the remark this answers is not in the document]
Comment (elvinpoole) [reply txpipe/ingest/ssi.py]: done

File: txpipe/ingest/ssi.py [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+        batch_size,
+        start,
+        end,
+    ):
+        """
+        Write the matched catalog for a single chunk
+        """
+        g = outfile[group]
+        for name, col in phot_group.items():
+            g[name][start:end] = col[ichunk * batch_size : (ichunk + 1) * batch_size][
```
[the remark this answers is not in the document]
Comment (elvinpoole) [reply txpipe/ingest/ssi.py:262]: done

File: txpipe/magnification.py (hunk 1 above)
Comment (elvinpoole) [reply txpipe/magnification.py:88]: done

File: txpipe/magnification.py (hunk 2 above)
Comment (elvinpoole) [reply txpipe/magnification.py:147]: done

File: txpipe/magnification.py (hunk 3 above)
Comment (elvinpoole) [reply txpipe/magnification.py]: yes good idea. i added an extra "bin" index that will contain the full sample `bin_all`

File: txpipe/magnification.py (hunk 4 above)
Comment (elvinpoole) [reply txpipe/magnification.py]: good catch. I've updated these to weighted counts

File: txpipe/ingest/ssi.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (joezuntz) [reply txpipe/ingest/ssi.py:41]: Makes sense.

File: txpipe/ingest/ssi.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (joezuntz) [reply txpipe/ingest/ssi.py:156]: No, that's fine if there's a reason.

**Comment by joezuntz** [comment]
@elvinpoole I think all your changes here are made now - is that right? If so I can merge master into it and then merge the PR.

**Comment by elvinpoole** [comment]
yes that's right. thank you

**Pull request merged**